core/bin/external_node/src/node_builder.rs

```diff
@@ -0,0 +1,511 @@
+        l1_batch_commitment_mode_validation::L1BatchCommitmentModeValidationLayer,
+        main_node_client::MainNodeClientLayer,
+        main_node_fee_params_fetcher::MainNodeFeeParamsFetcherLayer,
+        metadata_calculator::MetadataCalculatorLayer,
+        pools_layer::PoolsLayerBuilder,
+        postgres_metrics::PostgresMetricsLayer,
+        prometheus_exporter::PrometheusExporterLayer,
+        pruning::PruningLayer,
+        query_eth_client::QueryEthClientLayer,
+        sigint::SigintHandlerLayer,
+        state_keeper::{
+            external_io::ExternalIOLayer, main_batch_executor::MainBatchExecutorLayer,
+            output_handler::OutputHandlerLayer, StateKeeperLayer,
+        },
+        sync_state_updater::SyncStateUpdaterLayer,
+        tree_data_fetcher::TreeDataFetcherLayer,
+        validate_chain_ids::ValidateChainIdsLayer,
+        web3_api::{
+            caches::MempoolCacheLayer,
+            server::{Web3ServerLayer, Web3ServerOptionalConfig},
+            tree_api_client::TreeApiClientLayer,
+            tx_sender::{PostgresStorageCachesConfig, TxSenderLayer},
+            tx_sink::TxSinkLayer,
+        },
+    },
+    service::{ZkStackService, ZkStackServiceBuilder},
+};
+
+use crate::{
+    config::{self, ExternalNodeConfig},
+    Component,
+};
+
+/// Builder for the external node.
+#[derive(Debug)]
+pub(crate) struct ExternalNodeBuilder {
+    node: ZkStackServiceBuilder,
+    config: ExternalNodeConfig,
+}
+
+impl ExternalNodeBuilder {
+    pub fn new(config: ExternalNodeConfig) -> Self {
+        Self {
+            node: ZkStackServiceBuilder::new(),
+            config,
+        }
+    }
+
+    fn add_sigint_handler_layer(mut self) -> anyhow::Result<Self> {
+        self.node.add_layer(SigintHandlerLayer);
+        Ok(self)
+    }
+
+    fn add_pools_layer(mut self) -> anyhow::Result<Self> {
+        // Note: the EN config doesn't currently support specifying configuration for replicas,
+        // so we reuse the master configuration for that purpose.
+        // Settings unconditionally set to `None` are either not supported by the EN configuration layer
+        // or are not used in the context of the external node.
+        let config = PostgresConfig {
+            max_connections: Some(self.config.postgres.max_connections),
+            max_connections_master: Some(self.config.postgres.max_connections),
+            acquire_timeout_sec: None,
+            statement_timeout_sec: None,
+            long_connection_threshold_ms: None,
+            slow_query_threshold_ms: self
+                .config
+                .optional
+                .slow_query_threshold()
+                .map(|d| d.as_millis() as u64),
+            test_server_url: None,
+            test_prover_url: None,
+        };
+        let secrets = DatabaseSecrets {
+            server_url: Some(self.config.postgres.database_url()),
+            server_replica_url: Some(self.config.postgres.database_url()),
```

Comment:
do not understand why to set `server_replica_url` here

Reply:
Because api layer uses replica pool. In the old code, we could pass two copies of main pool to API, but the framework considers main pool and different pool as different resources -- so we configure the pools layer to provide both, currently pointing to the same (main) pool in fact.
In the future, we will probably want to support replica pool configuration for EN too, since we now support running in a distributed mode.